src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs

```diff
@@ -18,6 +18,14 @@ namespace System.Runtime.CompilerServices
         internal StringHandleOnStack([NotNull] ref string? s)
         {
             _ptr = Unsafe.AsPointer(ref s);
+
+            // The StringHandleOnStack type itself doesn't guarantee that 's' is
+            // non-null. But the typical usage pattern is that this handle is
+            // passed to a QCall routine, and that QCall routine will ensure that
+            // the captured reference is non-null. So even though the [NonNull]
+            // annotation here is a lie we'll keep it around for convenience.
```

Comment:
I suggest we actually be truthful here, remove the [NotNull], and change the call sites to use `!` where necessary.